substra/sdk/backends/local/backend.py

```diff
@@ -0,0 +1,523 @@
+            compute_plan_id = compute_plan.compute_plan_id
+            if len(in_traintuples) == 0:
+                rank = 0
+            else:
+                rank = 1 + max(
+                    [
+                        in_traintuple.rank
+                        for in_traintuple in in_traintuples
+                        if in_traintuple.compute_plan_id == compute_plan_id
+                    ]
+                )
+
+            #  Add to the compute plan
+            if compute_plan.traintuple_keys is None:
+                compute_plan.traintuple_keys = [key]
+            else:
+                compute_plan.traintuple_keys.append(key)
+            compute_plan.tuple_count += 1
+            compute_plan.status = models.Status.waiting
+
+        else:
+            compute_plan_id = ""
+            rank = 0
+
+        # create model
+        options = {}
+        traintuple = models.Traintuple(
+            key=key,
+            creator=_BACKEND_ID,
+            worker=_BACKEND_ID,
+            algo_key=spec.algo_key,
+            dataset={
+                "opener_hash": spec.data_manager_key,
+                "keys": spec.train_data_sample_keys,
+                "worker": _BACKEND_ID,
+            },
+            permissions={
+                "process": {"public": public, "authorized_ids": authorized_ids},
+            },
+            log="",
+            compute_plan_id=compute_plan_id,
+            rank=rank,
+            tag=spec.tag or "",
+            status=models.Status.waiting.value,
+            in_models=[
+                {
+                    "hash": in_traintuple.out_model.hash_,
+                    "storage_address": in_traintuple.out_model.storage_address,
+                }
+                for in_traintuple in in_traintuples
+            ],
+            metadata=spec.metadata if spec.metadata else dict(),
+            **options,
+        )
+
+        traintuple = self._db.add(traintuple, exist_ok)
+        self._worker.schedule_traintuple(traintuple)
+        return traintuple
+
+    def _add_testtuple(self, spec, exist_ok, spec_options=None):
+
+        # validation
+        objective = self._db.get(schemas.Type.Objective, spec.objective_key)
+        traintuple = self._db.get(schemas.Type.Traintuple, spec.traintuple_key)
+        assert traintuple.out_model is not None
+        if spec.data_manager_key is not None:
+            self._db.get(schemas.Type.Dataset, spec.data_manager_key)
+        if spec.test_data_sample_keys is not None:
+            same_data_manager = all(
+                [
+                    spec.data_manager_key
+                    in self._db.get(schemas.Type.DataSample, key).data_manager_keys
+                    for key in spec.test_data_sample_keys
+                ]
+            )
+            if not same_data_manager:
+                raise substra.exceptions.InvalidRequest(
+                    "dataSample do not belong to the same dataManager", 400
+                )
```

Comment:
It looks like this code is duplicated in this module, it could be moved to a separate function.